### DISCUSSION TITLE
Any documentation for the Mijia authorization protocol?

### DISCUSSION THREAD

**Discussion by swiergot** [pvvx/ATC_MiThermometer#433] [General]
Hi,

I'm trying to run the OTA flasher code in Node.js. It basically works but I'm observing a weird issue with Qingping Temp RH M.

When ran in the browser, messages "registration successful" and "login successful" are received from characteristic 10. But when I run it from Node.js, they come from characteristic 19. Oddly enough, both registration and login seem to be successful anyway, i.e. when "registration successful" comes from characteristic 19, I do the login procedure and it works ok. I have not progressed to actual flashing yet.

@pvvx any ideas why that happens?

Another issue is that quite often after sending 00000100, 15000000 and 000000030400 the sensor is not replying with anything. That happens in a browser too but rarely.

Which brings me to the ultimate question - is there any documentation describing this protocol?

Thanks in advance.

**Comment by pvvx** [comment]
This is not an Telink OTA protocol. This is a Mijia authorization protected by encryption.
In Xiaomi devices, authorization from Mijia is required to enable "Telink OTA" operation.
The codes used were obtained through "reverse engineering" - ATC1441 ported Java code from MIHome to js.

Additional Information:
https://github.com/MiEcosystem
https://iot.mi.com/new/doc/accesses/direct-access/overview


"Telink OTA" is fully published by Telink and is open source in the Telink SDK, available to everyone.

**Comment by pvvx** [comment]
> any ideas why that happens?

MiHome experiences the same instabilities :)
**You must press the button on the device as indicated in the instructions.**

**Comment by swiergot** [comment]
I do press the button, otherwise it rarely connects at all.

It's strange that the same code works differently in the browser and in Node.js. While the underlying BT implementation is probably different, messages sent to the sensor are the same. I have also tried rewriting the code to Python and it's even worse there, ie. if it somehow makes it past 000000030400 it always stops a bit later after transmitting the public key. The funny thing is, in ALL cases the BT communication goes via DBus and Bluez (I'm doing it on Linux). Although I have not analyzed if DBus messages are exactly the same.

Anyway, despite all the issues, flashing seems to be successful with Node.js, even if it's 7-8 times slower than in the browser. And to be precise, I **think** it is successful because the very last write fails for some reason, yet the sensor reboots into the new firmware. But I assume the received firmware's checksum is verified before the new firmware bank is marked as valid/next.
